src/st/cat/e_coffin.c

```diff
@@ -0,0 +1,172 @@
+static Point32 broken_coffin_parts_velocity[] = {
+    {FIX(-0.75), FIX(-2.5)}, {FIX(-0.625), FIX(-2.5)}, {FIX(-0.375), FIX(-1.5)},
+    {FIX(-1.5), FIX(-2.0)},  {FIX(-1.125), FIX(-2.0)}, {FIX(-1.0), FIX(-2.0)},
+};
+static s16 coffin_parts_bounce_posY[] = {10, 10, 8, 4, 0, 4};
+static s16 coffin_parts_bounce_velocity[] = {
+    FIX(7.0 / 32.0), FIX(17.0 / 128.0), FIX(5.0 / 32.0),
+    FIX(3.0 / 32.0), FIX(3.0 / 32.0),   FIX(1.0 / 8.0)};
+static s16 D_us_801816F0[] = {16, -64, -32, -64, -128, -48};
+
+// Params = 0 - spawns a blood skeleton
+// Params != 0 - spawns an item drop
+void EntityCoffin(Entity* self) {
+    Collider collider;
+    Entity* newEntity;
+    s32 i;
+    u8 params;
+    Entity* coffinSpawnEntity;
+    s16 posX;
+    s16 posY;
+
+    enum Step {
+        INIT = 0,
+        DESTROYED = 1,
+        SPAWN_SUB_ENTITIES = 2,
+        BROKEN_COFFIN_PARTS = 3,
+        COFFIN_PARTS_BOUNCE = 4,
+        DONE = 32
+    };
+
+    switch (self->step) {
+    case INIT:
+        InitializeEntity(g_EInitCoffin);
+        self->hitboxWidth = 8;
+        self->hitboxHeight = 0x14;
+        self->hitboxState = 2;
+        self->hitPoints = 2;
+        self->facingLeft = self->params >> 8;
+        self->params &= 0xFF;
+        self->zPriority = 0x70;
+        break;
+    case DESTROYED:
+        if (self->flags & FLAG_DEAD) {
+            g_api.PlaySfx(SFX_WALL_DEBRIS_B);
+            self->hitboxState = 0;
+            self->step++;
+        }
+        break;
+    case SPAWN_SUB_ENTITIES:
+        if (!AnimateEntity(anim_coffin_explode, self)) {
+            g_api.PlaySfx(SFX_FM_EXPLODE_B);
+
+            // Spawn the broken coffin parts
+            for (i = 0; i < 6; i++) {
+                newEntity = AllocEntity(&g_Entities[224], &g_Entities[256]);
+                if (newEntity != NULL) {
+                    CreateEntityFromEntity(E_ID(COFFIN), self, newEntity);
+                    newEntity->step = BROKEN_COFFIN_PARTS;
+                    newEntity->params = i;
+                    newEntity->facingLeft = self->facingLeft;
+                    newEntity->ext.coffin.playerOnLeft = GetSideToPlayer() & 1;
+                    newEntity->flags |= FLAG_UNK_2000;
+                }
+            }
+
+            newEntity = AllocEntity(&g_Entities[224], &g_Entities[256]);
+            if (newEntity != NULL) {
+                CreateEntityFromEntity(E_EXPLOSION, self, newEntity);
+                newEntity->params = ((self->zPriority + 2) << 8) + 0x11;
+            }
+
+            coffinSpawnEntity = AllocEntity(&g_Entities[160], &g_Entities[192]);
+            if (coffinSpawnEntity != NULL) {
+                if (!self->params) {
+                    CreateEntityFromCurrentEntity(
+                        E_ID(BLOOD_SKELETON), coffinSpawnEntity);
+                } else {
+                    CreateEntityFromCurrentEntity(
+                        E_HEART_DROP, coffinSpawnEntity);
```

Comment:
This spawns Monster Vial 3 out of these coffins, so I think `EntityHeartDrop` is probably inaccurately named.
I'm not certain exactly how it differs from `EntityPrizeDrop` though
https://github.com/Xeeynamo/sotn-decomp/blob/master/src/st/e_collect.h#L1471

Proposal to rename?